pkg/pillar/evetpm/tpm.go

```diff
@@ -389,8 +385,225 @@ func readDiskKey(log *base.LogObject) ([]byte, error) {
 	keyBytes, err := tpm2.NVReadEx(rw, TpmDiskKeyHdl,
 		tpm2.HandleOwner, EmptyPassword, 0)
 	if err != nil {
-		log.Errorf("NVReadEx failed: %v", err)
-		return nil, err
+		return nil, fmt.Errorf("NVReadEx failed: %v", err)
 	}
 	return keyBytes, nil
 }
+
+//FetchSealedVaultKey fetches Vault key sealed into TPM2.0
+func FetchSealedVaultKey() ([]byte, error) {
+
+	//gain some knowledge about existing environment
+	sealedKeyPresent := isSealedKeyPresent()
+	legacyKeyPresent := isLegacyKeyPresent()
+
+	if !sealedKeyPresent && !legacyKeyPresent {
+		//Fresh install, go for a new key
+		key, err := GetRandom(vaultKeyLength)
```

Comment:
Is this TPM Random function the recommended one for key generation? Or is it intended for strong random numbers for nonces etc? Need to be concerned about weak keys with certain crypto hence my question.

Reply:
We are using raw_key option to protect the encryption/decryption protector: 
https://github.com/google/fscrypt#using-a-raw-key-protector

fscrypt wants a random 32-byte binary value as the raw key value. It uses it as a seed
to further derive an AES key which will then protect the actual key used in encryption
and decryption of the data on disk.

To satisfy randomness, we are using TPM's native RNG component (preferred over /dev/urandom):
https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part1_Architecture_pub.pdf

The name "key" may be confusing here, as in it is not a key per se. It is a seed value AFAIK. But that is what
fscrypt uses, so I borrowed the same term for uniformity.

Reply:
Added this note as a comment